src/offsets.zig

```diff
@@ -1,280 +1,304 @@
-    while (i + start_index < end_index) {
-        std.debug.assert(source[i] != '\n');
+    const line_text = std.mem.sliceTo(text[line_start_index..], '\n');
+    const line_byte_length = getNCodeUnitByteCount(line_text, position.character, encoding);
 
-        const n = try std.unicode.utf8ByteSequenceLength(source[i]);
-        if (i + n >= source.len)
-            return error.CodepointTooLong;
-
-        const codepoint = try std.unicode.utf8Decode(source[i .. i + n]);
+    return line_start_index + line_byte_length;
+}
 
-        result += 1 + @as(usize, @boolToInt(codepoint >= 0x10000));
-        i += n;
-    }
-    return result;
+pub fn tokenToIndex(tree: Ast, token_index: Ast.TokenIndex) usize {
+    return tree.tokens.items(.start)[token_index];
 }
 
-pub const TokenLocation = struct {
-    line: usize,
-    column: usize,
-    offset: usize,
+pub fn tokenToLoc(tree: Ast, token_index: Ast.TokenIndex) Loc {
+    const start = tree.tokens.items(.start)[token_index];
+    const tag = tree.tokens.items(.tag)[token_index];
 
-    pub fn add(lhs: TokenLocation, rhs: TokenLocation) TokenLocation {
+    // Many tokens can be determined entirely by their tag.
+    if (tag.lexeme()) |lexeme| {
         return .{
-            .line = lhs.line + rhs.line,
-            .column = if (rhs.line == 0)
-                lhs.column + rhs.column
-            else
-                rhs.column,
-            .offset = rhs.offset,
+            .start = start,
+            .end = start + lexeme.len,
         };
     }
-};
-
-pub fn tokenRelativeLocation(tree: Ast, start_index: usize, token_start: usize, encoding: Encoding) !TokenLocation {
-    if (token_start < start_index)
-        return error.InvalidParams;
 
-    var loc = TokenLocation{
-        .line = 0,
-        .column = 0,
-        .offset = 0,
+    // For some tokens, re-tokenization is needed to find the end.
+    var tokenizer: std.zig.Tokenizer = .{
+        .buffer = tree.source,
+        .index = start,
+        .pending_invalid_token = null,
     };
 
-    const source = tree.source[start_index..];
-    var i: usize = 0;
-    while (i + start_index < token_start) {
-        const c = source[i];
-        if (c == '\n') {
-            loc.line += 1;
-            loc.column = 0;
-            i += 1;
-        } else {
-            if (encoding == .utf16) {
-                const n = try std.unicode.utf8ByteSequenceLength(c);
-                if (i + n >= source.len)
-                    return error.CodepointTooLong;
-
-                const codepoint = try std.unicode.utf8Decode(source[i .. i + n]);
-                loc.column += 1 + @as(usize, @boolToInt(codepoint >= 0x10000));
-                i += n;
-            } else {
-                loc.column += 1;
-                i += 1;
-            }
-        }
-    }
-    loc.offset = i + start_index;
-    return loc;
+    // Maybe combine multi-line tokens?
+    const token = tokenizer.next();
+    std.debug.assert(token.tag == tag);
```

Comment:
We you able to reproduce the example given in #381 with this PR? The assertion doesn't seem to fail for me.
Maybe there was a bug somewhere else that later causes this assertion to fail and my PR "accidentally" fixed it.

Reply:
Thanks for all the bug hunting! I've fixed the first two issues. About the issue in your latest message, I found out that it always fails because the token is `.eof` so I just added an exception which allows it (kind of hacky, I know ...).
I will need to do some further investigations on why this happens.
And also I would really like to avoid having to return an error, because the code in your example consists of valid tokens so why should `tokenToLoc` ever fail?

Reply:
`pending_invalid_token` deals with ill-formed utf8 encoded text and is only useful if you call `tokenizer.next` multiple times.
`tokenToLoc` only calls `tokenizer.next` once so its essentially unused and can be ignored.

Reply:
i wasn't correctly releasing a borrowed range in `completeFieldAccess` which inserted a null character at the cursor position. this should be fixed now.
note: for some reason github didn't publish this comment when posted it two days ago